File: PopulateNextRight.py
'''
Solution:
1.  In this question, instead of normal BFS which uses a queue, we can use the next pointers
    and update the node's next pointers at next level towards their nodes at their right using 
    a pointer pointing at current level's first node.
2.  If current node's left exists => point its next to node's right
    If current node's next exists => point its right to next node's left.
3.  The only difference between recursion and iteration is that recursion requires extra
    logN recursive stack space.

Time Complexity:    O(n) for both the cases.
Space Complexity:   O(h) for recursive solution and O(1) for iterative approach.

--- Passed all testcases successfully for both solutions on leetcode.xs
'''


# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next

class PopulateNextRightRecursion:
    def connect(self, root: 'Node') -> 'Node':
        
        #   base case
        if (root == None):
            return root
        
        #   if root's left exits => point its left next to its right
        if (root.left != None):
            root.left.next = root.right
            #   now if root's next exits => point its right's next to its next's left
            if (root.next != None):
                root.right.next = root.next.left
        
        #   recursion on left and right subtrees       
        self.connect(root.left)
        self.connect(root.right)
        
        #   return root
        return root

class PopulateNextRightIteration:
    def connect(self, root: 'Node') -> 'Node':
        
        #   edge case check
        if (root == None):
            return root
        
        #   initialize the first node of each level (now the main root)
        firstNodeOfLevel = root
        
        #   iterate until the level of the tree exists
        while (firstNodeOfLevel != None):
            
            #   iterate over the current level and assign next pointers of nodes at next level
            currentNode = firstNodeOfLevel
            while (currentNode != None):

                #   if current node's left exits => point its left next to its right
                if (currentNode.left != None):
                    currentNode.left.next = currentNode.right
                    #   now if current node's next exits => point its right's next to its next's left
                    if (currentNode.next != None):
                        currentNode.right.next = currentNode.next.left
                #   update current node to its next
                currentNode = currentNode.next
            
            #   update each level's first node once the level is complete  
            firstNodeOfLevel = firstNodeOfLevel.left
         
        #   return the root   
        return root

File: populateNextRight.java
//*116. Populating Next Right Pointers in Each Node
//Brut-force soultion
import java.util.*;

class Node {
    int val;
    Node next;
    Node left;
    Node right;
    Node() {}
    Node(int val) { this.val = val; }
    Node(int val, Node next, Node left, Node right) {
        this.val = val;
        this.next = next;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public Node connect(Node root) {
        Queue<Node> q = new LinkedList<>();
        if(root == null){
            return null;
        }
        q.add(root);
        while(!q.isEmpty()){
            int siz = q.size();
            for(int i = 0; i < siz; i++){
                Node current = q.poll();
                if(i != siz -1)
                current.next = q.peek();
            
            if(current.left != null){
                q.add(current.left);
            }
            if(current.right != null ){
                q.add(current.right);
            }
            }
        }
        return root;
    }
}


//Optimized soultion

class Solution1 {
    public Node connect(Node root) {
        if(root == null){
            return null;
        }
        Node level = root;
        while(level.left != null){
            Node current = level;
            while(current != null){
                current.left.next = current.right;
                if(current.next != null){
                    current.right.next = current.next.left;
                }
                current = current.next;
            }
            
            level= level.left;
        }
        
        return root;
    }
}